File: 99. Recover Binary Search Tree/99. Recover Binary Search Tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def recover_tree(root)
    @first = nil
    @second = nil
    @prev = nil
    
    inorder(root)
    @first.val, @second.val = @second.val, @first.val
end

def inorder(node)
    return if node.nil?
    
    inorder(node.left)
    
    if @prev && node.val < @prev.val
        @first = @prev if @first.nil?
        @second = node
    end
    @prev = node
    
    inorder(node.right)
end

File: 99. Recover Binary Search Tree/99. Recover Binary Search Tree.js
function recoverTree(root) {
    let first = null;
    let second = null;
    let prev = new TreeNode(-Infinity);

    // Helper function to perform inorder traversal
    function inorder(node) {
        if (node === null) return;

        inorder(node.left);

        if (node.val < prev.val) {
            if (first === null) {
                first = prev;
            }
            second = node;
        }
        prev = node;

        inorder(node.right);
    }

    inorder(root);

    // Swap values of the two nodes
    if (first !== null && second !== null) {
        let temp = first.val;
        first.val = second.val;
        second.val = temp;
    }
}

// Definition for a binary tree node
function TreeNode(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
}